// @duetds/tokens/lib/tokens.common.js
module.exports = {
  colorWarning: "rgb(247, 178, 40)",
  colorGrayDarker: "rgb(101, 119, 135)",
  colorGrayLighter: "rgb(245, 248, 250)",
  colorCategoryHome: "rgb(211, 36, 131)",
  colorPrimary: "rgb(0, 119, 179)",
  colorCategoryVehicle: "rgb(54, 74, 129)",
  colorSuccess: "rgb(0, 135, 90)",
  colorCategoryFamily: "rgb(11, 114, 136)",
  colorPrimaryLightest: "rgb(243, 249, 252)",
  colorPrimaryLighter: "rgb(230, 242, 248)",
  colorBrandBlue: "rgb(0, 161, 212)",
  colorPrimaryDark: "rgb(0, 77, 128)",
  colorGrayLightest: "rgb(255, 255, 255)",
  colorCategoryFinance: "rgb(122, 1, 196)",
  colorGrayDark: "rgb(144, 149, 153)",
  colorGrayLight: "rgb(225, 227, 230)",
  colorSecondary: "rgb(0, 41, 77)",
  colorCategoryPet: "rgb(168, 1, 78)",
  colorPrimaryLight: "rgb(205, 229, 241)",
  colorCategoryTravel: "rgb(213, 66, 3)",
  colorGray: "rgb(207, 210, 212)",
  colorDanger: "rgb(222, 35, 98)",
  colorGrayDarkerTurva: "rgb(68, 68, 69)",
  colorGrayLighterTurva: "rgb(245, 245, 247)",
  colorCategoryHomeTurva: "rgb(161, 10, 78)",
  colorPrimaryTurva: "rgb(198, 12, 48)",
  colorCategoryVehicleTurva: "rgb(23, 28, 58)",
  colorSuccessTurva: "rgb(8, 135, 78)",
  colorPrimaryLightestTurva: "rgb(252, 243, 244)",
  colorCategoryFamilyTurva: "rgb(199, 12, 112)",
  colorCategoryUnionTurva: "rgb(169, 69, 124)",
  colorPrimaryLighterTurva: "rgb(249, 230, 234)",
  colorPrimaryDarkTurva: "rgb(148, 9, 37)",
  colorGrayLightestTurva: "rgb(255, 255, 255)",
  colorGrayDarkTurva: "rgb(116, 116, 117)",
  colorGrayLightTurva: "rgb(228, 228, 230)",
  colorBrandGrayTurva: "rgb(178, 180, 179)",
  colorSecondaryTurva: "rgb(23, 28, 58)",
  colorPrimaryLightTurva: "rgb(244, 207, 214)",
  colorCategoryTravelTurva: "rgb(9, 129, 148)",
  colorGrayTurva: "rgb(207, 207, 209)",
  colorDangerTurva: "rgb(224, 42, 13)",
  colorWarningTurva: "rgb(250, 164, 15)",
  colorData07: "rgb(192, 101, 247)",
  colorData08: "rgb(0, 105, 191)",
  colorData09: "rgb(19, 120, 53)",
  colorData10: "rgb(179, 60, 24)",
  colorData11: "rgb(13, 163, 166)",
  colorData01: "rgb(122, 1, 196)",
  colorData12: "rgb(11, 114, 136)",
  colorData02: "rgb(0, 155, 204)",
  colorData13: "rgb(179, 137, 24)",
  colorData03: "rgb(27, 171, 75)",
  colorData14: "rgb(128, 98, 17)",
  colorData04: "rgb(255, 86, 34)",
  colorData15: "rgb(0, 41, 77)",
  colorData05: "rgb(245, 76, 154)",
  colorData06: "rgb(194, 21, 101)",
  fontWeightExtraBold: "800",
  letterSpacingHeading: "-0.01rem",
  fontFamilyHeadingTurva: "'turva-sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
  fontFamilyTextTurva: "'turva-sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
  fontFamilyHeading: "'localtapiola-sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
  fontFamilyText: "'localtapiola-sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
  fontWeightBold: "700",
  fontWeightNormal: "400",
  fontWeightSemiBold: "600",
  fontSizeXxxxLarge: "4.5rem",
  fontSizeXxxLarge: "3rem",
  fontSizeXxLarge: "2.25rem",
  fontSizeXLarge: "1.5rem",
  fontSizeLarge: "1.25rem",
  fontSizeMedium: "1rem",
  fontSizeSmall: "0.875rem",
  fontSizeXSmall: "0.75rem",
  radiusSharp: "0",
  radiusDefault: "4px",
  radiusCircle: "50%",
  radiusPill: "20rem",
  shadowTooltipTurva: "0 2px 10px 0 rgba(117, 117, 117, 0.13)",
  shadowDefault: "0 2px 6px 0 rgba(0, 41, 77, 0.07)",
  shadowTooltip: "0 2px 10px 0 rgba(0, 41, 77, 0.07)",
  shadowCardTurva: "0 2px 6px 0 rgba(117, 117, 117, 0.13), 0 -1px 0 0 rgba(0, 0, 0, 0.09), -1px 0 0 0 rgba(0, 0, 0, 0.07), 1px 0 0 0 rgba(0, 0, 0, 0.07), 0 1px 0 0 rgba(0, 0, 0, 0.07)",
  shadowModal: "0 2px 20px 0 rgba(0, 0, 0, 0.2)",
  shadowCard: "0 2px 6px 0 rgba(0, 41, 77, 0.07), 0 -1px 0 0 rgba(0, 0, 0, 0.09), -1px 0 0 0 rgba(0, 0, 0, 0.07), 1px 0 0 0 rgba(0, 0, 0, 0.07), 0 1px 0 0 rgba(0, 0, 0, 0.07)",
  shadowHoverTurva: "0 2px 10px 0 rgba(117, 117, 117, 0.15)",
  shadowHover: "0 2px 10px 0 rgba(0, 41, 77, 0.1)",
  shadowDefaultTurva: "0 2px 6px 0 rgba(117, 117, 117, 0.13)",
  spaceXxSmall: "4px",
  spaceLarge: "20px",
  spaceXSmall: "8px",
  spaceXxxSmall: "2px",
  spaceSmall: "12px",
  spaceMedium: "16px",
  spaceXxLarge: "36px",
  spaceXxxxLarge: "72px",
  spaceXxxxxLarge: "94px",
  spaceXxxLarge: "48px",
  spaceXLarge: "28px",
  sizeIconXxLarge: "48px",
  sizeFormBorder: "1px",
  sizeStepMedium: "40px",
  sizeButtonMinWidth: "8rem",
  sizeHeader: "4rem",
  sizeIconXxxLarge: "72px",
  sizeIconXLarge: "36px",
  sizeIconXxSmall: "10px",
  sizeStepSmall: "30px",
  sizeNavigation: "3rem",
  sizeIconLarge: "30px",
  sizeLayoutMaxWidth: "1110px",
  sizeTappableSquare: "48px",
  sizeIconXSmall: "16px",
  sizeIconXxxSmall: "7px",
  sizeButtonBorder: "2px",
  sizeButtonTinyBorder: "1px",
  sizeIconMedium: "24px",
  sizeIconSmall: "20px",
  opacity85: "0.85",
  opacity75: "0.75",
  opacity50: "0.50",
  opacity30: "0.30",
  zIndexMaskButtons: "250",
  zIndexDropdown: "600",
  zIndexSpinner: "800",
  zIndexDefault: "1",
  zIndexToast: "500",
  zIndexPopup: "950",
  zIndexModal: "900",
  zIndexSticky: "300",
  zIndexMasked: "100",
  zIndexDeep: "-999999",
  zIndexOverlay: "700",
  zIndexMask: "200",
  zIndexHeader: "400",
  lineHeightMedium: "1.5",
  lineHeightSmall: "1.25",
  lineHeightXSmall: "1.1",
  transitionQuickly: "300ms ease",
  transitionSlowly: "600ms ease",
  mediaQueryXxxLarge: "'(min-width: 106.25em)'",
  mediaQueryXLarge: "'(min-width: 64.0625em)'",
  mediaQueryMedium: "'(min-width: 48em)'",
  mediaQueryLarge: "'(min-width: 62em)'",
  mediaQueryXxLarge: "'(min-width: 76.25em)'",
  mediaQueryXSmall: "'(max-width: 35.9375em)'",
  mediaQueryXxxSmall: "'(max-width: 20.5em)'",
  mediaQuerySmall: "'(min-width: 36em)'",
  mediaQueryXxSmall: "'(max-width: 22.5em)'",
};
