// @duetds/components/lib/esm/duet-empty-state.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host } from './index-81967d12.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';

const duetEmptyStateCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%;height:100%}.duet-empty-state{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:100%;padding:48px;color:#00294d;text-align:center;background:#f5f8fa;border-radius:4px}.duet-empty-state.duet-theme-turva{color:#171c3a;background:#f5f5f7}";

const DuetEmptyState = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Theme of the empty state component.
     */
    this.theme = "";
    /**
     * Icon shown in the empty state component.
     */
    this.icon = "messaging-question";
    /**
     * The size of the empty state message.
     */
    this.size = "medium";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("div", { class: {
        "duet-empty-state": true,
        "duet-theme-turva": this.theme === "turva",
        "duet-is-small": this.size === "small",
      } }, h("duet-icon", { color: "currentColor", name: this.icon, size: this.size === "small" ? "large" : "x-large" }), h("div", null, h("slot", null)))));
  }
};
DuetEmptyState.style = duetEmptyStateCss;

export { DuetEmptyState as duet_empty_state };
