// @duetds/components/lib/esm/duet-action-button.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-81967d12.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';
import { c as createID } from './create-id-981107da.js';
import { d as debounce } from './js-utils-fa98b8cd.js';
import { i as isKeyboardClick } from './keyboard-utils-88ea1007.js';
import { g as getLocaleString } from './language-utils-bf701d37.js';
import { c as capitalizeFirstLetter, g as getColorByName, s as shadeRGBColor } from './token-utils-ebf797ab.js';
import './string-utils-5954f458.js';
import './tokens.module-385c4cf8.js';

const duetActionButtonCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:inline-block;margin:0 auto}button.duet-action-button{white-space:nowrap}button.duet-action-button{display:inline-flex;align-items:center;justify-content:center;width:32px;height:32px;min-height:auto !important;padding:0 !important;margin:0 0 0 8px !important;color:#00294d;cursor:pointer;background:#f5f8fa;border-radius:50%;transition:background-color 300ms ease !important}.duet-theme-turva button.duet-action-button{color:#171c3a;background:#f5f5f7}button.duet-action-button duet-icon{display:block}@media (max-width: 35.9375em){button.duet-action-button{width:32px;height:32px}}button.duet-action-button:not(:disabled):hover{background:#e9ecee}.duet-theme-turva button.duet-action-button:not(:disabled):hover{background:#e9e9eb}button.duet-action-button:focus{outline:0;box-shadow:0 0 0 2px #0077b3 !important}.duet-theme-turva button.duet-action-button:focus{box-shadow:0 0 0 2px #171c3a !important}button.duet-action-button:active:focus{background:#dddfe1 !important;box-shadow:none}.duet-theme-turva button.duet-action-button:active:focus{background:#ddddde !important}button.duet-action-button:disabled{color:#909599;cursor:default;background:#f5f8fa;opacity:0.75}.duet-theme-turva button.duet-action-button:disabled{color:#747475;background:#f5f5f7}";

const DuetActionButton = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.duetActionEvent = createEvent(this, "duetActionEvent", 6);
    this.listId = createID("duet-action-button");
    this.numbersStore = "";
    /**
     * Theme of the pagination.
     */
    this.theme = "";
    /**
     * Name of the icon to display
     */
    this.iconName = "action-arrow-left-small";
    /**
     * Color of the icon to display
     */
    this.iconColor = "currentColor";
    /**
     * Size of the icon
     */
    this.iconSize = "xx-small";
    /**
     * Name of action being called
     */
    this.actionName = "default-action";
    /**
     * Name of id of action that was called being called
     */
    this.actionId = undefined;
    /**
     * Name of id of action that was called being called
     */
    this.actionMeta = undefined;
    /**
     * Disables the button
     */
    this.disabled = false;
    /**
     * Custom color to be used for the icon, as a design token entered in camelCase
     * or kebab-case. Example: "color-primary". This property can also be set to
     * "currentColor" which forces the icon to use the CSS text color of parent
     * element instead. Useful when you want to control the color in stylesheet
     * instead.
     */
    this.color = "";
    /**
     * Custom color to be used for a background, as a design token entered
     * in camelCase or kebab-case. Using this option the icon will always be resized
     * to $size-icon-medium on mobile viewports. Example: "color-primary".
     */
    this.background = "";
    /**
     * Title of the action button (shows as DOM tooltip when hovering)
     * if set to true, this will try to add a best-guess title based on action-name
     * if set to a string the string is used
     * if set to false nothing is set
     */
    this.accessibleTitle = true;
    /**
     * Described By id
     */
    this.accessibleDescribedBy = undefined;
    /**
     * Details of the component
     */
    this.accessibleDetails = undefined;
    /**
     * String of id's that indicate alternative labels elements
     */
    this.accessibleLabelledBy = undefined;
    /**
     * Aria description the button
     */
    this.accessibleDescription = undefined;
    /**
     * Default accessible labels
     */
    this.accessibleLabelsDefaults = {
      en: {
        label: "Click to activate action",
      },
      fi: {
        label: "Klikkaa toiminnon aktivointiin",
      },
      sv: {
        label: "Klicka för att aktivera åtgärd",
      },
    };
    /**
     * Default accessible labels
     */
    this.accessibleLabels = getLocaleString(this.accessibleLabelsDefaults);
    /**
     * Used to indicate which dom element with ID this element controls
     */
    this.accessibleControls = "";
    /**
     * Private methods.
     */
    //helper to emit relevant page events
    this.emitPageEvent = debounce(e => {
      this.duetActionEvent.emit({
        component: "duet-action-button",
        originalEvent: e,
        action: this.actionId,
        name: this.actionName,
        meta: this.actionMeta,
        id: this.listId,
      });
    });
  }
  //handle a11y keyboard navigation events
  async onKeyboardDown(e) {
    if (isKeyboardClick(e)) {
      this.emitPageEvent(e);
    }
  }
  //return readable title for the action-button
  getTitle() {
    let titleString = "";
    if (typeof this.accessibleTitle === "string") {
      titleString = this.accessibleTitle;
    }
    if (this.accessibleTitle === true) {
      titleString = capitalizeFirstLetter(`${this.actionName}`);
    }
    return titleString;
  }
  //set update current when mouse click on page number
  async onMouseHandler(e) {
    e.preventDefault();
    this.duetActionEvent.emit({
      component: "duet-action-button",
      originalEvent: e,
      action: this.actionName,
      id: this.actionId,
      meta: this.actionMeta,
    });
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const color = this.color === "currentColor" ? "currentColor" : getColorByName(this.color, this.theme);
    const backgroundColor = getColorByName(this.background, this.theme);
    const backgroundHoverColor = backgroundColor ? shadeRGBColor(backgroundColor, -0.1) : backgroundColor;
    const backgroundStyles = backgroundColor
      ? `
          button.duet-action-button {
            background: ${backgroundColor} !important;
          }
          button.duet-action-button:not(:disabled):hover {
            background: ${backgroundHoverColor} !important;
          }
        `
      : "";
    const styles = {
      color,
      "border-color": getColorByName(this.iconColor, this.theme),
      "border-width": "1px",
      "border-style": "solid",
    };
    return (h(Host, { onKeyDown: e => this.onKeyboardDown(e) }, h("style", { innerHTML: backgroundStyles }), h("button", { part: "button", style: styles, disabled: this.disabled, id: this.listId, class: {
        "duet-action-button": true,
        "duet-action-button--disabled": this.disabled,
        [`duet-action-button--${this.theme}`]: this.theme !== "",
      }, type: "button", onClick: e => this.onMouseHandler(e), "aria-label": this.accessibleLabels.label, "aria-labelledby": this.accessibleLabelledBy, "aria-description": this.accessibleDescription, "aria-describedby": this.accessibleDescribedBy, "aria-details": this.accessibleDetails, "aria-controls": this.accessibleControls, title: this.getTitle() }, h("duet-icon", { part: "icon", name: this.iconName, color: this.iconColor, margin: "none", size: this.iconSize }), h("duet-visually-hidden", null, h("slot", null)))));
  }
  get element() { return getElement(this); }
};
DuetActionButton.style = duetActionButtonCss;

export { DuetActionButton as duet_action_button };
