// @duetds/components/lib/esm/duet-upload-aria-status.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host } from './index-81967d12.js';

const DuetUploadAriaStatus = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * State() variables
     */
    this.statusMessage = null;
    /**
     * Status labels used in the component
     */
    this.statusMessageLabel = null;
    /**
     * Adds accessible label that is read aloud
     */
    this.accessibleAriaLive = "polite";
    /**
     * valid file amount
     */
    this.valid = undefined;
    /**
     * total files
     */
    this.total = undefined;
    /**
     * invalid file amount
     */
    this.invalid = undefined;
    /**
     * valid file amount
     */
    this.inprogress = undefined;
    /**
     * Private functions
     */
    this.replacePlaceHolders = (statusString) => {
      const addFiles = (amount) => {
        return amount > 1 ? `${amount} ${this.statusMessageLabel.files}` : `${amount} ${this.statusMessageLabel.file}`;
      };
      return statusString
        .replace("{filesUploaded}", addFiles(this.valid))
        .replace("{filesInProgress}", addFiles(this.inprogress))
        .replace("{filesTotal}", addFiles(this.total))
        .replace("{filesWithErrors}", addFiles(this.invalid));
    };
    this.getStatusMessage = () => {
      const { done, inProgress, doneWithErrors, inProgressWithErrors } = this.statusMessageLabel;
      const hasErrors = this.invalid > 0;
      const isInProgress = this.inprogress > 0;
      const isInProgressWithErrors = this.inprogress > 0 && hasErrors;
      const isDone = this.valid === this.total && !isInProgress && !hasErrors;
      const isDoneWithErrors = this.valid + this.invalid === this.total && !isInProgress;
      if (isDone) {
        this.statusMessage = this.replacePlaceHolders(done);
        return;
      }
      else if (isInProgressWithErrors) {
        this.statusMessage = this.replacePlaceHolders(inProgressWithErrors);
        return;
      }
      else if (isInProgress) {
        this.statusMessage = this.replacePlaceHolders(inProgress);
        return;
      }
      else if (isDoneWithErrors) {
        this.statusMessage = this.replacePlaceHolders(doneWithErrors);
        return;
      }
      return;
    };
  }
  watchHandler(newValue, oldValue) {
    if (newValue !== oldValue) {
      this.getStatusMessage();
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { role: "status", "aria-live": "assertive", "aria-atomic": "true", "aria-relevant": "all" }, h("duet-visually-hidden", null, this.statusMessage)));
  }
  static get watchers() { return {
    "valid": ["watchHandler"],
    "invalid": ["watchHandler"],
    "inprogress": ["watchHandler"],
    "total": ["watchHandler"]
  }; }
};

export { DuetUploadAriaStatus as duet_upload_aria_status };
