// @duetds/components/lib/cjs/duet-notification_2.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-d380d72b.js');
const themeableComponent = require('./themeable-component-fdeaef9f.js');
const languageUtils = require('./language-utils-4f62a7db.js');
const actionArrowDownSmall = require('./action-arrow-down-small-ef77a2d1.js');
require('./string-utils-5cadd058.js');

const duetNotificationCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.duet-notification{display:block;padding:16px;color:#00294d;cursor:pointer;border-bottom:1px solid #e1e3e6;transition:300ms ease}:host(:last-of-type) .duet-notification{border-bottom:0}.duet-notification.duet-theme-turva{color:#171c3a;border-bottom-color:#e4e4e6}.duet-notification.duet-notification-highlight{font-weight:600;background:#f3f9fc}.duet-notification.duet-notification-highlight.duet-theme-turva{background:#fcf3f4}@media (min-width: 62em){.duet-notification:hover{background:#f3f9fc}.duet-notification:hover.duet-theme-turva{background:#fcf3f4}}.duet-notification a{text-decoration:none}.duet-notification a:focus{outline:0}:host(.user-is-tabbing) .duet-notification:focus-within{z-index:1;box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px #0077b3}:host(.user-is-tabbing) .duet-notification:focus-within.duet-theme-turva{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px #171c3a}.duet-notification-container{position:relative;display:flex}.duet-notification-badge-container{position:relative;width:16px}.duet-notification-badge{position:absolute;top:6px;left:-2px;width:8px;height:8px;background:#0077b3;border-radius:50%}.duet-theme-turva .duet-notification-badge{background:#c60c30}.duet-notification-content{flex:1}.duet-notification-date{position:absolute;top:2px;right:0;width:96px;font-size:0.75rem;font-weight:600;line-height:1.25;color:#657787;text-align:right;text-transform:uppercase}@media (max-width: 35.9375em){.duet-notification-date{width:auto}}.duet-theme-turva .duet-notification-date{color:#444445}::slotted(duet-heading){padding-right:80px !important}::slotted(duet-paragraph){margin-bottom:2px !important}";

const DuetNotification = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * The currently active language.
     * Used for announcing unread notifications to screen readers.
     * @deprecated this is now handled via the html lang tag, and is no longer used - kept to avoid breaking changes and ease unit testing
     * @default "fi"
     */
    this.language = languageUtils.getLanguage();
    /**
     * Property to change accessibleLabelUnread defaults on the component.
     * normally you would handle these strings on an application level and override @accessibleLabelUnread when needed
     */
    this.accessibleLabelUnreadDefaults = {
      en: "Unread",
      fi: "Lukematon",
      sv: "Oläst",
    };
    /**
     * Adds accessible label for tooltip that is shown over unread counter
     * @default { en: "Unread", fi: "Lukematon", sv: "Oläst" }
     */
    this.accessibleLabelUnread = languageUtils.getLocaleString(this.accessibleLabelUnreadDefaults, this.language);
    /**
     * The url to open on notification click.
     */
    this.url = "#";
    /**
     * The date of the notification.
     */
    this.date = "";
    /**
     * Should the notification be visually highlighted?
     */
    this.highlight = false;
    /**
     * Should the link open in a new window?
     */
    this.external = false;
    /**
     * Theme of the notification.
     */
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    themeableComponent.inheritGlobalTheme(this);
  }
  /**
   * Move focus to the notification's link element.
   * Use this method instead of the native `focus()` method.
   * @param options
   */
  async setFocus(options) {
    this.linkElement.focus(options);
  }
  render() {
    return (index.h("div", { role: "listitem", class: {
        "duet-notification": true,
        "duet-notification-highlight": this.highlight,
        "duet-theme-turva": this.theme === "turva",
      } }, index.h("a", { ref: element => (this.linkElement = element), href: this.url, target: this.external ? "_blank" : "_self" }, index.h("div", { class: "duet-notification-container" }, index.h("div", { class: "duet-notification-badge-container" }, this.highlight && (index.h("div", { class: "duet-notification-badge" }, index.h("duet-visually-hidden", null, this.accessibleLabelUnread, ":")))), index.h("div", { class: "duet-notification-content" }, index.h("slot", null)), index.h("div", { class: "duet-notification-date" }, this.date)))));
  }
  get element() { return index.getElement(this); }
};
DuetNotification.style = duetNotificationCss;

const duetNotificationDrawerCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.duet-drawer{position:relative}@media (min-width: 62em){.duet-drawer{display:inline-block}}.duet-drawer-toggle{position:relative;display:flex;align-items:center;width:100%;height:48px;padding:0 20px;overflow:hidden;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:1rem;font-weight:400;font-variant-numeric:tabular-nums;line-height:48px;color:#00294d;text-align:left;text-decoration:none;text-overflow:ellipsis;white-space:nowrap;cursor:pointer;background-color:white;border-bottom:1px solid #e1e3e6;border-radius:0;-webkit-appearance:none;appearance:none}@media (min-width: 62em){.duet-drawer-toggle{z-index:200;width:auto;max-width:200px;padding:0 30px 0 12px;overflow:hidden;font-size:0.875rem;line-height:4rem;border:0;border-radius:4px;transition:border 300ms ease}}.duet-theme-turva .duet-drawer-toggle{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:#171c3a;background-color:white;border-color:#e4e4e6}.duet-drawer-toggle:active{opacity:0.75}.duet-drawer-toggle:focus{outline:0}:host(.user-is-tabbing) .duet-drawer-toggle:focus{box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px #0077b3}:host(.user-is-tabbing) .duet-theme-turva .duet-drawer-toggle:focus{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px #171c3a}.duet-drawer-toggle:hover{color:#004d80}.duet-theme-turva .duet-drawer-toggle:hover{color:#444445}.duet-drawer-toggle::after{display:none !important}@media (min-width: 62em){.duet-drawer-toggle::after{position:absolute;top:auto;bottom:0;left:0;display:block !important;width:100%;height:3px;content:\"\";background:#0077b3;transition:300ms ease;transform:translateY(3px)}.duet-theme-turva .duet-drawer-toggle::after{background:#c60c30}}.duet-drawer-toggle[aria-expanded=true],.duet-drawer-toggle[aria-expanded=true]:hover{color:#004d80;border-bottom:1px solid #f5f8fa;border-radius:0}.duet-theme-turva .duet-drawer-toggle[aria-expanded=true],.duet-theme-turva .duet-drawer-toggle[aria-expanded=true]:hover{color:#444445;border-color:#f5f5f7}@media (min-width: 62em){.duet-drawer-toggle[aria-expanded=true],.duet-drawer-toggle[aria-expanded=true]:hover{border:0}.duet-drawer-toggle[aria-expanded=true]::after,.duet-drawer-toggle[aria-expanded=true]:hover::after{left:0;width:100%;transform:translateY(0)}}.duet-drawer-icon{position:relative;display:inline-flex;min-width:12px;margin-right:12px;font-size:0;line-height:normal;vertical-align:baseline}@media (min-width: 62em){.duet-drawer-icon{margin-right:8px}}.duet-drawer-label{line-height:1}.duet-drawer-caret{position:absolute;top:50%;right:20px;display:flex;width:10px;height:10px;margin-left:8px;line-height:normal;pointer-events:none;transition:300ms ease;transform:translateY(-50%);transform-origin:50% 50%}@media (min-width: 62em){.duet-drawer-caret{right:16px;width:7px;height:7px}}.duet-drawer-caret duet-icon{width:10px}.duet-drawer-toggle[aria-expanded=true] .duet-drawer-caret{transform:translateY(-50%) rotate(180deg)}.duet-drawer-badge{position:absolute;top:0;right:0;display:inline-block;width:6px;height:6px;background:#f7b228;border-radius:50%;box-shadow:0 0 0 1px #c18b1f}.duet-theme-turva .duet-drawer-badge{background:#faa40f}.duet-drawer-content{display:none;background-color:white;border-bottom:1px solid #e1e3e6}.duet-drawer-content:focus{outline:0}.duet-theme-turva .duet-drawer-content{background-color:white;border-color:#e1e3e6}@media (min-width: 62em){.duet-drawer-content{position:absolute;top:100%;z-index:600;display:block;width:450px;max-height:420px;padding-left:0;overflow-y:auto;visibility:hidden;background-color:white;border:1px solid #e1e3e6;border-radius:4px;border-top-left-radius:0;border-top-right-radius:0;box-shadow:0 4px 10px 0 rgba(0, 41, 77, 0.15);opacity:0;transition:300ms ease;transform:scale(0.95) translateZ(0) translateY(-10px);transform-origin:50% 0;scrollbar-width:none}.duet-drawer-content::-webkit-scrollbar{display:none}.duet-theme-turva .duet-drawer-content{background-color:white}.duet-drawer-content.duet-drawer-left{right:0}.duet-drawer-content.duet-drawer-right{left:0}}.duet-drawer-content.duet-drawer-open{display:block}@media (min-width: 62em){.duet-drawer-content.duet-drawer-open{visibility:visible;opacity:1;transform:scale(1) translateZ(0) translateY(0)}}";

const ESC_KEY = 27;
const DuetNotificationDrawer = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * State() variables
     * Inlined decorator, alphabetical order.
     */
    this.isOpen = false;
    /**
     * Icon to display to the left of the drawer button.
     */
    this.icon = "";
    /**
     * Label for the drawer button.
     */
    this.label = "Label";
    /**
     * The currently active language.
     * Used to determine the message displayed when there are no notifications.
     */
    this.language = languageUtils.getLanguage();
    /**
     * Defaults for accessibleLabelNotifications
     * @default {en: "No notifications", fi: "Ei ilmoituksia", sv: "Inga notifikationer"}
     */
    this.accessibleLabelNotificationsDefaults = {
      en: "No notifications",
      fi: "Ei ilmoituksia",
      sv: "Inga notifikationer",
    };
    /**
     * Adds accessible label for tooltip that is shown for notifications
     * @default {en: "No notifications", fi: "Ei ilmoituksia", sv: "Inga notifikationer"}
     */
    this.accessibleLabelNotifications = languageUtils.getLocaleString(this.accessibleLabelNotificationsDefaults);
    /**
     * Show a badge that indicates something has changed.
     */
    this.badge = false;
    /**
     * Theme of the notification drawer.
     */
    this.theme = "";
    /**
     * Which direction should the drawer open in?
     */
    this.direction = "left";
    /**
     * Component event handling.
     */
    this.handleKeyDown = (ev) => {
      if (this.isOpen && ev.keyCode === ESC_KEY) {
        this.close();
      }
    };
    this.toggleMenu = (e) => {
      // this is needed to prevent the main menu closing on mobile
      e.stopPropagation();
      this.isOpen ? this.close() : this.open();
    };
    this.updateNotificationState = () => {
      this.hasNotifications = this.element.children.length > 0;
    };
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    themeableComponent.inheritGlobalTheme(this);
    this.updateNotificationState();
  }
  connectedCallback() {
    this.observer = new MutationObserver(this.updateNotificationState);
    this.observer.observe(this.element, { childList: true });
  }
  disconnectedCallback() {
    this.observer.disconnect();
    this.observer = null;
  }
  handleClickFocusOutside(e) {
    const target = e.target;
    if (this.isOpen && !this.element.contains(target)) {
      this.close(false);
    }
  }
  /**
   * Move focus to the menu's button. Use instead of the native `focus()` method
   * @param options
   */
  async setFocus(options) {
    this.drawerButtonEl.focus(options);
  }
  /**
   * Open the menu
   */
  async open() {
    const header = this.element.closest("duet-header");
    if (header) {
      header.closeMenus();
    }
    this.isOpen = true;
    this.drawerContentEl.scrollTop = 0;
  }
  /**
   * Close the menu
   * @param moveFocusToButton should focus be moved to the menu's button on close?
   */
  async close(moveFocusToButton = true) {
    this.isOpen = false;
    if (moveFocusToButton) {
      this.setFocus({ preventScroll: true });
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (index.h("div", { class: {
        "duet-drawer": true,
        "duet-theme-turva": this.theme === "turva",
      }, onKeyDown: this.handleKeyDown }, index.h("button", { ref: el => (this.drawerButtonEl = el), type: "button", class: "duet-drawer-toggle", id: "duet-drawer-toggle", "aria-controls": "duet-drawer-content", "aria-expanded": this.isOpen ? "true" : "false", onClick: this.toggleMenu }, this.icon && (index.h("div", { class: "duet-drawer-icon" }, index.h("duet-icon", { margin: "none", size: "x-small", name: this.icon, color: "currentColor" }), this.badge && index.h("div", { class: "duet-drawer-badge" }))), index.h("span", { class: "duet-drawer-label" }, this.label), index.h("div", { class: "duet-drawer-caret" }, index.h("duet-icon", { margin: "none", size: "auto", icon: actionArrowDownSmall.actionArrowDownSmall.svg, color: "currentColor" }))), index.h("div", { id: "duet-drawer-content", class: {
        "duet-drawer-content": true,
        "duet-drawer-right": this.direction === "right",
        "duet-drawer-left": this.direction === "left",
        "duet-drawer-open": this.isOpen,
      }, ref: element => (this.drawerContentEl = element) }, this.hasNotifications ? (index.h("div", { role: "list" }, index.h("slot", null))) : (index.h("duet-empty-state", { size: "small" }, index.h("duet-paragraph", { size: "small", margin: "none" }, this.accessibleLabelNotifications))))));
  }
  get element() { return index.getElement(this); }
};
DuetNotificationDrawer.style = duetNotificationDrawerCss;

exports.duet_notification = DuetNotification;
exports.duet_notification_drawer = DuetNotificationDrawer;
