// @duetds/components/lib/dist-custom-elements/duet-range-stepper.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h } from './p-df05a117.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { c as createID } from './p-c1325e35.js';
import { h as isArrowLeftKey, e as isArrowUpKey, n as isMinusKey, l as isArrowKey, j as isArrowRightKey, f as isArrowDownKey, o as isPlusKey } from './p-d8549efe.js';
import { g as getLocaleString } from './p-ede854e1.js';
import { d as defineCustomElement$4 } from './p-f06badc1.js';
import { d as defineCustomElement$3 } from './p-983633e4.js';
import { d as defineCustomElement$2 } from './p-14799264.js';

const duetRangeStepperCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:inline-block;margin:0 auto}.duet-range-stepper{display:flex;flex-wrap:wrap;gap:0;align-items:baseline;justify-content:center;min-width:30px;min-height:30px;font-weight:600;text-decoration:none;border-left-width:0}.duet-range-stepper{white-space:nowrap}.duet-range-stepper .duet-range-step-counter{font-weight:600;color:#00294d}.duet-range-stepper ::part(button){border:0}";

const DuetRangeStepper$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.duetRangeStepUpdate = createEvent(this, "duetRangeStepUpdate", 7);
    this.leftBn = createID("duetRangeStepper-left");
    this.rightBn = createID("duetRangeStepper-right");
    /**
     * Theme of the component.
     */
    this.theme = "";
    /**
     * Start value
     * @default 1
     */
    this.stepIndex = 1;
    /**
     * How big is the step
     * @default 50
     */
    this.stepSize = 50;
    /**
     * Total of
     * @default undefined
     */
    this.total = 1000;
    /**
     * Default range stepper labels
     */
    this.ariaLabelsDefaults = {
      en: {
        next_increment: "Next range increment",
        previous_increment: "Previous range increment",
        x_of_y: "{0}–{1} of {2}",
      },
      fi: {
        next_increment: "Seuraava välillä",
        previous_increment: "Edellinen välillä",
        x_of_y: "{0}–{1} / {2}",
      },
      sv: {
        next_increment: "Nästa intervall",
        previous_increment: "Föregående intervall",
        x_of_y: "{0}–{1} av {2}",
      },
    };
    /**
     * Default pagination labels
     */
    this.ariaLabels = getLocaleString(this.ariaLabelsDefaults);
    /**
     * Used to indicate which dom element with ID this element controls
     */
    this.ariaControls = "";
    this.handleClick = (e, direction) => {
      e.preventDefault();
      this.stepIndex += direction;
      this.clampStepIndex();
      this.handleStepClick(e);
    };
  }
  watchStepIndexStateHandler(newValue) {
    if (this.element.shadowRoot.childElementCount > 0) {
      if (newValue <= 1) {
        this.leftBnRef.setAttribute("disabled", "");
        this.rightBnRef.focus();
      }
      else {
        this.leftBnRef.removeAttribute("disabled");
      }
      if (newValue >= this.total / this.stepSize) {
        this.rightBnRef.setAttribute("disabled", "");
        this.leftBnRef.focus();
      }
      else {
        this.rightBnRef.removeAttribute("disabled");
      }
    }
  }
  watchStepSizeStateHandler(newValue, oldValue) {
    if (newValue !== oldValue) {
      if (this.stepIndex * newValue > this.total) {
        this.stepIndex = Math.ceil(this.total / newValue);
      }
    }
  }
  handleStepClick(e) {
    const from = this.stepIndex * this.stepSize - this.stepSize;
    this.duetRangeStepUpdate.emit({
      originalEvent: e,
      component: "duet-range-stepper",
      from: from,
      to: this.getTo(),
      index: this.stepIndex,
    });
  }
  componentWillLoad() {
    inheritGlobalTheme(this);
    this.clampStepIndex();
  }
  componentDidLoad() {
    this.watchStepIndexStateHandler(this.stepIndex);
  }
  clampStepIndex() {
    if (this.stepIndex < 1) {
      this.stepIndex = 1;
    }
    else if (this.stepIndex * this.stepSize > this.total) {
      this.stepIndex = Math.ceil(this.total / this.stepSize);
    }
  }
  getTo() {
    const to = this.stepIndex * this.stepSize;
    const result = to > this.total ? this.total : to;
    return result;
  }
  generateStepSize() {
    // this function uses this.stepIndex and this.range to generate a range of number like 1-100
    // this is used to generate the step size
    let start = Math.floor((this.stepIndex - 1) * this.stepSize);
    //easy fix for starting position
    if (start === 0) {
      start = 1;
    }
    const end = Math.floor(this.stepIndex * this.stepSize);
    return this.ariaLabels.x_of_y
      .replace("{0}", String(start))
      .replace("{1}", String(end))
      .replace("{2}", String(this.total));
  }
  //handle a11y keyboard navigation events
  async onKeyboardDown(e) {
    // handle a11y based keyboard navigation for left/right/up/down
    let next = this.stepIndex;
    let dir = 1;
    if (isArrowLeftKey(e) || isArrowUpKey(e) || isMinusKey(e)) {
      next--;
      dir = -1;
      this.leftBnRef.focus();
    }
    if (isArrowRightKey(e) || isArrowDownKey(e) || isPlusKey(e)) {
      next++;
      dir = 1;
    }
    if (isArrowKey(e)) {
      if (next > 0 && next * this.stepSize <= this.total) {
        this.handleClick(e, dir);
      }
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h("div", { class: {
        "duet-range-stepper": true,
        "duet-theme-turva": this.theme === "turva",
      }, onKeyDown: e => this.onKeyboardDown(e), "aria-controls": this.ariaControls }, h("span", { class: "duet-range-step-counter", part: "range-numbers" }, this.generateStepSize()), h("duet-action-button", { class: "duet-range-step-button duet-range-step-minus", onClick: e => this.handleClick(e, -1), id: this.leftBn, "aria-label": this.ariaLabels.previous_increment, ref: el => (this.leftBnRef = el), part: "navigation-arrow", iconName: "action-arrow-left-small" }), h("duet-action-button", { class: "duet-range-step-button duet-range-step-plus", onClick: e => this.handleClick(e, 1), "aria-label": this.ariaLabels.next_increment, id: this.rightBn, ref: el => (this.rightBnRef = el), part: "navigation-arrow", iconName: "action-arrow-right-small" })));
  }
  get element() { return this; }
  static get watchers() { return {
    "stepIndex": ["watchStepIndexStateHandler"],
    "stepSize": ["watchStepSizeStateHandler"]
  }; }
  static get style() { return duetRangeStepperCss; }
}, [1, "duet-range-stepper", {
    "theme": [1025],
    "stepIndex": [1026, "step-index"],
    "stepSize": [2, "step-size"],
    "total": [2],
    "ariaLabelsDefaults": [16],
    "ariaLabels": [16],
    "ariaControls": [1, "accessible-controls"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-range-stepper", "duet-action-button", "duet-icon", "duet-visually-hidden"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-range-stepper":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetRangeStepper$1);
      }
      break;
    case "duet-action-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "duet-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "duet-visually-hidden":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetRangeStepper = DuetRangeStepper$1;
const defineCustomElement = defineCustomElement$1;

export { DuetRangeStepper, defineCustomElement };
