// @duetds/components/lib/dist-custom-elements/duet-choice-group.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, b as Host } from './p-df05a117.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { f as findCheckedOption, w as watchForOptions } from './p-12721178.js';
import { d as defineCustomElement$6 } from './p-a10df7a8.js';
import { d as defineCustomElement$5 } from './p-c2260d50.js';
import { d as defineCustomElement$4 } from './p-983633e4.js';
import { d as defineCustomElement$3 } from './p-5c30c1e6.js';
import { d as defineCustomElement$2 } from './p-14799264.js';

const duetChoiceGroupCss = "*.sc-duet-choice-group,*.sc-duet-choice-group::after,*.sc-duet-choice-group::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.sc-duet-choice-group-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.horizontal.sc-duet-choice-group-h .duet-choice-group.sc-duet-choice-group{display:block}@media (min-width: 48em){.horizontal.sc-duet-choice-group-h .duet-choice-group.sc-duet-choice-group{display:flex;margin-bottom:0}}.horizontal.sc-duet-choice-group-h .duet-choice-group.duet-no-stacking.sc-duet-choice-group{display:flex}.duet-choice-group.sc-duet-choice-group{position:relative;width:100%;margin-bottom:-12px}";

const duetChoiceSelector = ":scope duet-choice:not(:scope duet-choice duet-choice-group duet-choice):not(:scope duet-choice duet-fieldset duet-choice)";
const DuetChoiceGroup$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.choiceGroupName = "duet-choice-group";
    /**
     * Direction of the choice group.
     */
    this.direction = "vertical";
    /**
     * Controls the margin of the component.
     */
    this.margin = "auto";
    /**
     * Theme of the choice group.
     */
    this.theme = "";
    /**
     * Display choice group in error state along with an error message.
     */
    this.error = "";
    /**
     * Legend displayed for the choice buttons in this group.
     */
    this.label = "Label";
    /**
     * Visually hide the label, but still show it to screen readers.
     */
    this.labelHidden = false;
    /**
     * Additional caption to show inside the label of the choice group.
     */
    this.caption = undefined;
    /**
     * Tooltip to display next to the label of the choice group.
     */
    this.tooltip = undefined;
    /**
     * With direction setting you can force the tooltip to always open towards left
     * or right instead of automatically determining the direction.
     */
    this.tooltipDirection = "auto";
    /**
     * Enable or disable the automatic responsive behaviour of the choice group
     * component when horizontal setting is used. Setting this option to "true"
     * makes sure that contents are stacked vertically on mobile.
     */
    this.responsive = false;
    this.getChoices = () => {
      return Array.from(this.element.querySelectorAll(duetChoiceSelector));
    };
    this.isMyChoice = (choice) => {
      return this.getChoices().includes(choice);
    };
    /**
     * Component event handling.
     */
    this.onSelect = (ev) => {
      if (ev.detail.component !== "duet-choice") {
        return;
      }
      const selectedChoice = ev.target;
      if (selectedChoice && this.isMyChoice(selectedChoice)) {
        this.value = selectedChoice.value;
      }
      // Prevent events from bubbling up the DOM tree in case
      // this group is a sub-group of another choice group.
      ev.cancelBubble = true;
    };
  }
  /**
   * Component lifecycle events.
   */
  async connectedCallback() {
    const el = this.element;
    if (this.value === undefined) {
      const choice = findCheckedOption(el, duetChoiceSelector);
      if (choice !== undefined) {
        await choice.componentOnReady();
        if (this.value === undefined) {
          this.value = choice.value;
        }
      }
    }
    this.mutationO = watchForOptions(el, duetChoiceSelector, newOption => {
      if (newOption !== undefined) {
        newOption.componentOnReady().then(() => {
          if (this.isMyChoice(newOption)) {
            this.value = newOption.value;
          }
        });
      }
      else {
        this.updateChoices();
      }
    });
    this.updateChoices();
  }
  disconnectedCallback() {
    if (this.mutationO) {
      this.mutationO.disconnect();
      this.mutationO = undefined;
    }
  }
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * Local methods
   */
  async updateChoices() {
    /**
     * Make sure we get all choices first so values are up to date prior
     * to caching the choice group value
     */
    const choices = this.getChoices();
    const { value } = this;
    let hasChecked = false;
    // Walk the DOM in reverse order, since the last selected one wins
    for (const choice of choices) {
      choice.name = this.name;
      choice.groupDisabled = this.disabled;
      choice.groupDirection = this.direction;
      choice.groupResponsive = this.responsive;
      if (!hasChecked && choice.value === value) {
        // Correct value for this choice, but this choice isn't checked yet
        // and we haven't found a checked yet
        hasChecked = true;
        choice.checked = true;
      }
      else {
        // This choice doesn't have the correct value or the choice group has been already checked
        choice.checked = false;
      }
    }
    // Reset value if
    if (!hasChecked) {
      this.value = undefined;
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { onDuetChange: this.onSelect, class: { horizontal: this.direction === "horizontal" } }, h("duet-fieldset", { label: this.label, labelHidden: this.labelHidden, caption: this.caption, error: this.error, margin: this.margin, theme: this.theme }, this.tooltip && (h("duet-tooltip", { slot: "tooltip", direction: this.tooltipDirection, accessibleInputLabel: this.label }, this.tooltip)), h("div", { class: { [this.choiceGroupName]: true, "duet-no-stacking": !this.responsive } }, h("slot", null)))));
  }
  get element() { return this; }
  static get watchers() { return {
    "value": ["updateChoices"],
    "disabled": ["updateChoices"],
    "name": ["updateChoices"],
    "direction": ["updateChoices"],
    "responsive": ["updateChoices"]
  }; }
  static get style() { return duetChoiceGroupCss; }
}, [6, "duet-choice-group", {
    "direction": [1],
    "margin": [1],
    "theme": [1025],
    "error": [1],
    "name": [1],
    "label": [1],
    "labelHidden": [4, "label-hidden"],
    "caption": [1],
    "tooltip": [1],
    "tooltipDirection": [1, "tooltip-direction"],
    "responsive": [4],
    "value": [1537],
    "disabled": [4]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-choice-group", "duet-caption", "duet-fieldset", "duet-icon", "duet-tooltip", "duet-visually-hidden"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-choice-group":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetChoiceGroup$1);
      }
      break;
    case "duet-caption":
      if (!customElements.get(tagName)) {
        defineCustomElement$6();
      }
      break;
    case "duet-fieldset":
      if (!customElements.get(tagName)) {
        defineCustomElement$5();
      }
      break;
    case "duet-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "duet-tooltip":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "duet-visually-hidden":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetChoiceGroup = DuetChoiceGroup$1;
const defineCustomElement = defineCustomElement$1;

export { DuetChoiceGroup, defineCustomElement };
