// @duetds/components/lib/collection/components/duet-badge/duet-badge.js
/*!
 * Built with Duet Design System
 */
import { Component, h, Host, Prop } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetBadge {
  constructor() {
    /**
     * Theme of the badge.
     */
    this.theme = "";
    /**
     * Style variation of the alert.
     */
    this.variation = "default";
    /**
     * Controls the margin of the component.
     */
    this.margin = "auto";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } },
      h("div", { class: {
          "duet-badge": true,
          [this.variation]: true,
          "duet-theme-turva": this.theme === "turva",
        } },
        h("slot", null))));
  }
  static get is() { return "duet-badge"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["duet-badge.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["duet-badge.css"]
  }; }
  static get properties() { return {
    "theme": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "DuetTheme",
        "resolved": "\"\" | \"default\" | \"turva\"",
        "references": {
          "DuetTheme": {
            "location": "import",
            "path": "../../common-types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Theme of the badge."
      },
      "attribute": "theme",
      "reflect": false,
      "defaultValue": "\"\""
    },
    "variation": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "DuetStatus",
        "resolved": "\"danger\" | \"default\" | \"success\" | \"warning\"",
        "references": {
          "DuetStatus": {
            "location": "import",
            "path": "../../common-types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Style variation of the alert."
      },
      "attribute": "variation",
      "reflect": false,
      "defaultValue": "\"default\""
    },
    "margin": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "DuetMargin",
        "resolved": "\"auto\" | \"none\"",
        "references": {
          "DuetMargin": {
            "location": "import",
            "path": "../../common-types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Controls the margin of the component."
      },
      "attribute": "margin",
      "reflect": false,
      "defaultValue": "\"auto\""
    }
  }; }
}
